design/design-generic-references.md

```diff
@@ -0,0 +1,211 @@
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
+In order to enable such use cases, we need a generic version of the cross-resource
+references where you can specify all metadata required to find the referenced
+resource as well as details about how to extract the value you need and where to
+put it in the referencer object.
+
+## Proposal
+
+All managed resources will have a top-level spec field that lists the generic
+references. They will be resolved in order and failure of any will cause
+reconciliation to fail. Inspired from the [latest developments in API conventions](https://github.com/kubernetes/community/pull/5748)
+regarding multi-kind references, the API will look like the following:
+```yaml
+# A generic reference to another managed resource.
+spec:
+  externalValues:
+  - fromObject:
+      group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fieldPath: spec.forProvider.cidrBlock
+    toFieldPath: spec.cidrBlock
+```
+```yaml
+# A generic reference to a namespaced Kubernetes resource.
+spec:
+  externalValues:
+  - fromObject:
+      version: v1
+      resource: configmaps
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+    toFieldPath: spec.forProvider.region
+```
+
+The syntax of `fieldPath` and `toFieldPath` fields will be similar to what we
+use in Composition, in line with [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/744e270/contributors/devel/sig-architecture/api-conventions.md#selecting-fields).
+
+The provider pods will resolve the references via managed reconciler, hence they
+will need at least read-only permissions for any resource that's been referenced
+by an instance. Currently, every provider [has read access](https://github.com/crossplane/crossplane/blob/d8f57a8/internal/controller/rbac/provider/roles/roles.go#L66)
+to all of its own kinds, `Secret`s, `ConfigMap`s, `Lease`s and `Event`s in all
+namespaces. In order to reference resources other than the ones listed, users will
+have to create necessary RBAC resources manually to grant the permissions.
+
+### Implementation
+
+The new field will exist in all managed resources and the cross-resource reference
+fields will stay as is and we'll continue adding them wherever possible. They will
+still be the main way to refer to another resource. The generic references will
+help in cases where that's not possible or not implemented yet.
```

Comment:
I like the idea of maintaining the existing references for the time being in order to avoid a breaking change, but if we adopt this new model my preference would be to deprecate the old explicitly-coded method. That way there is one clear way for users to define references.

Reply:
Though I agree having to add strongly typed references has been a challenge for writing/generating new managed resources, I actually lean towards keeping the existing references as first class. Because its API is much simpler for the cases where we know all the information we need during build time, i.e. references to other managed resources in the same provider, which is the case most of the time in my experience. Though this may change in the future if generated providers choose not to implement strongly-typed references for some reason, so we can come back and change the statement to make generic patches first class in such cases.

Reply:
> I agree that existing references have a simpler API but I would also prefer to have one way of doing referencing and deprecate the old way unless we are losing some functionality. 

In terms of functionality, we need to have more advanced transforms to cover some of the referencing mechanisms. For example, when resource-a requires ARN of resource-b, that ARN is not always available and you have to construct it. We do these constructions in Go today with the strongly typed references and users need to give only the name of the target resource. When you'd like to do the same with generic references, you need to give bunch of identification information as well as a set of transforms, so it gets pretty ugly in terms of UX and we can keep that from happening in build time. I think for users the flow will be that they will look at CRD schema and likely see the ref/selector field, if not, they know they can patch anything on the resource and will do that. But if they have to write generic ref patch for 7 fields on the same CR, which could be base template in the composition, the error surface increases and you get to know it after applying

Reply:
I remain convinced that we should not maintain multiple different ways of cross-resource references. As @turkenh says, having multiple ways to do one thing makes things complicated for our users. FWIW I also prefer the current reference implementation, which I feel is a more intuitive API than what we're likely to land on here. The issue is how manual that API is to support.

Is it possible for us to keep working on this design until we have one that we feel it can handle all the cases covered by the current XRR implementation in a reasonably elegant fashion?

Reply:
> Is it possible for us to keep working on this design until we have one that we feel it can handle all the cases covered by the current XRR implementation in a reasonably elegant fashion?

We can keep discussing the API shape but I'm not sure how much simplification we can ever get. Once the identification happens in runtime, the search space is `all kinds * all resources * all fields * field transformation` and we need to ask for information for each of those variables whereas current XRR search space is only `all resources` because kind and field and field transformations needed are known in build time, which makes it very easy to use.

> FWIW I also prefer the current reference implementation, which I feel is a more intuitive API than what we're likely to land on here. The issue is how manual that API is to support.

The manual support of that API could actually be automated with the code generation as long as this identification information that is to be written in build time is provided to the generator by the developer. In fact, I had started [working on that for ACK](https://github.com/aws-controllers-k8s/code-generator/pull/11) a while ago but postponed due to other reasons related to how ACK type generation works but I believe it's doable.

In fact, while our initial use case is generic referencing I believe we should think of this mechanism something similar to composition patches rather than a new referencing mechanism. Even though XRR was our initial use case, there is https://github.com/crossplane/crossplane/issues/2099 and possibly other use cases like implicit dependency management between managed resources that this feature will solve. I can see for example new types of patches here, like `fromVault` or `fromHTTP` that's used to fetch information from other sources and patch the resource.